File: backend/app/routers/groups.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.auth import get_current_user, hash_password, verify_password
from app.database import get_db
from app.models import Group, Membership, Transaction, TransactionType, User
from app.schemas import (
    GroupCreate,
    GroupJoinRequest,
    GroupLookup,
    GroupOut,
    LeaderboardEntry,
    MembershipOut,
    MyGroupSummary,
)
from app.services import valuation

router = APIRouter(prefix="/groups", tags=["groups"])


def group_out(group: Group) -> GroupOut:
    out = GroupOut.model_validate(group)
    out.member_count = len(group.memberships)
    out.has_password = group.join_password_hash is not None
    return out


def require_membership(group: Group, user: User) -> Membership:
    membership = next((m for m in group.memberships if m.user_id == user.id), None)
    if not membership:
        raise HTTPException(403, "You are not a member of this group")
    return membership


def _create_membership(db: Session, user: User, group: Group) -> Membership:
    membership = Membership(
        user_id=user.id,
        group_id=group.id,
        cash_balance=group.initial_cash,
        last_allowance_at=datetime.now(timezone.utc),
    )
    db.add(membership)
    db.flush()
    db.add(
        Transaction(
            membership_id=membership.id,
            type=TransactionType.INITIAL_DEPOSIT,
            amount=group.initial_cash,
        )
    )
    return membership


def _leaderboard_entries(db: Session, group: Group) -> list[LeaderboardEntry]:
    entries = []
    for membership in group.memberships:
        holdings_value, _ = valuation.get_holdings_value(db, membership)
        total = membership.cash_balance + holdings_value
        deposited = group.initial_cash + sum(
            (t.amount for t in membership.transactions if t.type == TransactionType.ALLOWANCE),
            start=0,
        )
        profit = total - deposited
        entries.append(
            LeaderboardEntry(
                membership_id=membership.id,
                user=membership.user,
                cash_balance=membership.cash_balance,
                holdings_value=holdings_value,
                total_value=total,
                profit=profit,
                profit_pct=round(profit / deposited * 100, 2) if deposited else 0,
            )
        )
    entries.sort(key=lambda e: e.total_value, reverse=True)
    return entries


@router.post("", response_model=GroupOut, status_code=201)
def create_group(
    payload: GroupCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    group = Group(
        name=payload.name,
        owner_id=current_user.id,
        initial_cash=payload.initial_cash,
        monthly_allowance=payload.monthly_allowance,
        join_password_hash=hash_password(payload.password) if payload.password else None,
    )
    db.add(group)
    db.flush()
    _create_membership(db, current_user, group)  # owner joins their own group
    db.commit()
    db.refresh(group)
    return group_out(group)


@router.get("/mine", response_model=list[MyGroupSummary])
def my_groups(
    db: Session = Depends(get_db), current_user: User = Depends(get_current_user)
):
    """Dashboard data: every group I'm in, with my standing in each."""
    memberships = db.scalars(
        select(Membership).where(Membership.user_id == current_user.id)
    ).all()

    summaries = []
    for membership in memberships:
        group = membership.group
        entries = _leaderboard_entries(db, group)
        mine = next(e for e in entries if e.membership_id == membership.id)
        rank = entries.index(mine) + 1
        summaries.append(
            MyGroupSummary(
                group=group_out(group),
                membership_id=membership.id,
                cash_balance=mine.cash_balance,
                total_value=mine.total_value,
                profit=mine.profit,
                rank=rank,
                is_owner=group.owner_id == current_user.id,
            )
        )
    summaries.sort(key=lambda s: s.group.created_at)
    return summaries


@router.get("/lookup/{invite_code}", response_model=GroupLookup)
def lookup_group(
    invite_code: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Preview a group by invite code before joining."""
    group = db.scalars(
        select(Group).where(Group.invite_code == invite_code.upper().strip())
    ).first()
    if not group:
        raise HTTPException(404, "No group found with that invite code")
    return GroupLookup(
        name=group.name,
        member_count=len(group.memberships),
        initial_cash=group.initial_cash,
        monthly_allowance=group.monthly_allowance,
        requires_password=group.join_password_hash is not None,
        already_member=any(m.user_id == current_user.id for m in group.memberships),
    )


@router.post("/join", response_model=MembershipOut, status_code=201)
def join_group(
    payload: GroupJoinRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    group = db.scalars(
        select(Group).where(Group.invite_code == payload.invite_code.upper().strip())
    ).first()
    if not group:
        raise HTTPException(404, "No group found with that invite code")
    if group.join_password_hash is not None:
        if not payload.password or not verify_password(
            payload.password, group.join_password_hash
        ):
            raise HTTPException(403, "Wrong group password")
    if any(m.user_id == current_user.id for m in group.memberships):
        raise HTTPException(409, "You are already a member of this group")

    membership = _create_membership(db, current_user, group)
    db.commit()
    db.refresh(membership)
    return membership


@router.get("/{group_id}", response_model=GroupOut)
def get_group(
    group_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    group = db.get(Group, group_id)
    if not group:
        raise HTTPException(404, "Group not found")
    require_membership(group, current_user)
    return group_out(group)


@router.get("/{group_id}/leaderboard", response_model=list[LeaderboardEntry])
def leaderboard(
    group_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    group = db.get(Group, group_id)
    if not group:
        raise HTTPException(404, "Group not found")
    require_membership(group, current_user)
    return _leaderboard_entries(db, group)

File: backend/src/routes/groups.js
import { Hono } from "hono";
import { hashPassword, requireUser, serializeUser, verifyPassword } from "../auth.js";
import { membershipHoldings, valueHoldings } from "../valuation.js";
import {
  ApiError,
  generateInviteCode,
  isoTimestamp,
  positiveNumber,
  r2,
  requireFields,
} from "../util.js";

const router = new Hono();
router.use("*", requireUser);

export const serializeGroup = (group, memberCount) => ({
  id: group.id,
  name: group.name,
  owner_id: group.owner_id,
  initial_cash: group.initial_cash,
  monthly_allowance: group.monthly_allowance,
  invite_code: group.invite_code,
  created_at: isoTimestamp(group.created_at),
  member_count: memberCount,
  has_password: group.join_password_hash !== null,
});

export const serializeMembership = (m) => ({
  id: m.id,
  user_id: m.user_id,
  group_id: m.group_id,
  cash_balance: m.cash_balance,
  joined_at: isoTimestamp(m.joined_at),
});

async function getGroup(env, groupId) {
  const group = await env.DB.prepare("SELECT * FROM groups WHERE id = ?").bind(groupId).first();
  if (!group) throw new ApiError(404, "Group not found");
  return group;
}

async function groupMembers(env, groupId) {
  const { results } = await env.DB.prepare(
    `SELECT m.*, u.id AS u_id, u.username, u.display_name, u.email, u.picture,
            u.created_at AS u_created_at
     FROM memberships m JOIN users u ON u.id = m.user_id WHERE m.group_id = ?`
  ).bind(groupId).all();
  return results;
}

async function memberCount(env, groupId) {
  const row = await env.DB.prepare(
    "SELECT COUNT(*) AS n FROM memberships WHERE group_id = ?"
  ).bind(groupId).first();
  return row.n;
}

async function createMembership(env, userId, group) {
  const membership = await env.DB.prepare(
    `INSERT INTO memberships (user_id, group_id, cash_balance, last_allowance_at)
     VALUES (?, ?, ?, ?) RETURNING *`
  ).bind(userId, group.id, group.initial_cash, Date.now()).first();
  await env.DB.prepare(
    "INSERT INTO transactions (membership_id, type, amount) VALUES (?, 'INITIAL_DEPOSIT', ?)"
  ).bind(membership.id, group.initial_cash).run();
  return membership;
}

async function leaderboardEntries(env, group) {
  const members = await groupMembers(env, group.id);
  const entries = [];
  for (const member of members) {
    const holdings = await membershipHoldings(env, member.id);
    const { total: holdingsValue } = await valueHoldings(env, holdings);
    const totalValue = r2(member.cash_balance + holdingsValue);
    const allowances = await env.DB.prepare(
      `SELECT COALESCE(SUM(amount), 0) AS total FROM transactions
       WHERE membership_id = ? AND type = 'ALLOWANCE'`
    ).bind(member.id).first();
    const deposited = r2(group.initial_cash + allowances.total);
    const profit = r2(totalValue - deposited);
    entries.push({
      membership_id: member.id,
      user: serializeUser({
        id: member.u_id,
        username: member.username,
        display_name: member.display_name,
        email: member.email,
        picture: member.picture,
        created_at: member.u_created_at,
      }),
      cash_balance: member.cash_balance,
      holdings_value: holdingsValue,
      total_value: totalValue,
      profit,
      profit_pct: deposited > 0 ? r2((profit / deposited) * 100) : 0,
    });
  }
  entries.sort((a, b) => b.total_value - a.total_value);
  return entries;
}

async function requireMembership(env, group, user) {
  const membership = await env.DB.prepare(
    "SELECT * FROM memberships WHERE group_id = ? AND user_id = ?"
  ).bind(group.id, user.id).first();
  if (!membership) throw new ApiError(403, "You are not a member of this group");
  return membership;
}

router.post("/", async (c) => {
  const body = await c.req.json().catch(() => ({}));
  requireFields(body, ["name", "initial_cash"]);
  if (String(body.name).trim().length < 2) throw new ApiError(422, "Group name is too short");
  const initialCash = positiveNumber(body.initial_cash, "initial_cash");
  const monthlyAllowance = Number(body.monthly_allowance ?? 0);
  if (!Number.isFinite(monthlyAllowance) || monthlyAllowance < 0) {
    throw new ApiError(422, "monthly_allowance must be zero or positive");
  }
  if (body.password && String(body.password).length < 4) {
    throw new ApiError(422, "Group password must be at least 4 characters");
  }

  const group = await c.env.DB.prepare(
    `INSERT INTO groups (name, owner_id, initial_cash, monthly_allowance, invite_code, join_password_hash)
     VALUES (?, ?, ?, ?, ?, ?) RETURNING *`
  )
    .bind(
      String(body.name).trim(),
      c.var.user.id,
      r2(initialCash),
      r2(monthlyAllowance),
      generateInviteCode(),
      body.password ? await hashPassword(body.password) : null
    )
    .first();
  await createMembership(c.env, c.var.user.id, group); // owner joins their own group
  return c.json(serializeGroup(group, 1), 201);
});

router.get("/mine", async (c) => {
  const { results: memberships } = await c.env.DB.prepare(
    "SELECT * FROM memberships WHERE user_id = ?"
  ).bind(c.var.user.id).all();

  const summaries = [];
  for (const membership of memberships) {
    const group = await getGroup(c.env, membership.group_id);
    const entries = await leaderboardEntries(c.env, group);
    const index = entries.findIndex((e) => e.membership_id === membership.id);
    const mine = entries[index];
    summaries.push({
      group: serializeGroup(group, entries.length),
      membership_id: membership.id,
      cash_balance: mine.cash_balance,
      total_value: mine.total_value,
      profit: mine.profit,
      rank: index + 1,
      is_owner: group.owner_id === c.var.user.id,
    });
  }
  summaries.sort((a, b) => (a.group.created_at < b.group.created_at ? -1 : 1));
  return c.json(summaries);
});

router.get("/lookup/:code", async (c) => {
  const code = c.req.param("code").toUpperCase().trim();
  const group = await c.env.DB.prepare("SELECT * FROM groups WHERE invite_code = ?")
    .bind(code)
    .first();
  if (!group) throw new ApiError(404, "No group found with that invite code");
  const already = await c.env.DB.prepare(
    "SELECT id FROM memberships WHERE group_id = ? AND user_id = ?"
  ).bind(group.id, c.var.user.id).first();
  return c.json({
    name: group.name,
    member_count: await memberCount(c.env, group.id),
    initial_cash: group.initial_cash,
    monthly_allowance: group.monthly_allowance,
    requires_password: group.join_password_hash !== null,
    already_member: Boolean(already),
  });
});

router.post("/join", async (c) => {
  const body = await c.req.json().catch(() => ({}));
  requireFields(body, ["invite_code"]);
  const group = await c.env.DB.prepare("SELECT * FROM groups WHERE invite_code = ?")
    .bind(String(body.invite_code).toUpperCase().trim())
    .first();
  if (!group) throw new ApiError(404, "No group found with that invite code");
  if (group.join_password_hash !== null) {
    const ok = body.password && (await verifyPassword(body.password, group.join_password_hash));
    if (!ok) throw new ApiError(403, "Wrong group password");
  }
  const existing = await c.env.DB.prepare(
    "SELECT id FROM memberships WHERE group_id = ? AND user_id = ?"
  ).bind(group.id, c.var.user.id).first();
  if (existing) throw new ApiError(409, "You are already a member of this group");

  const membership = await createMembership(c.env, c.var.user.id, group);
  return c.json(serializeMembership(membership), 201);
});

router.get("/:id", async (c) => {
  const group = await getGroup(c.env, Number(c.req.param("id")));
  await requireMembership(c.env, group, c.var.user);
  return c.json(serializeGroup(group, await memberCount(c.env, group.id)));
});

router.get("/:id/leaderboard", async (c) => {
  const group = await getGroup(c.env, Number(c.req.param("id")));
  await requireMembership(c.env, group, c.var.user);
  return c.json(await leaderboardEntries(c.env, group));
});

export default router;
